### PR TITLE
Added support for GCS

### LINKED ISSUE #70: Support GCS as object storage provider
Part of elastisys/ck8s-cluster#100

To be able to support GCP we need to be able to support to use a separate objectstorage than s3 (gcs)

## Acceptance criteria 
* Make it possible to choose what object storage provider to use
* Make sure that all apps to support gcs
* Configure the apps correctly depending on what object storage provider you choose

### PR DESCRIPTION
**Pull request by Xartos**
**What this PR does / why we need it**:

For the GCP support, we need to support GCS as object storage provider.

**Which issue this PR fixes** *(use the format `fixes #<issue number>(, fixes #<issue_number>, ...)` to automatically close the issue when PR gets merged)*: fixes #

fixes #70 

**Public facing documentation PR** *(if applicable)*

**Special notes for reviewer**:

**Checklist:**

- [x] Added relevant notes to [WIP-CHANGELOG.md](https://github.com/elastisys/compliantkubernetes-apps/blob/main/WIP-CHANGELOG.md)
- [x] Proper commit message prefix on all commits
- [ ] Updated the [public facing documentation](https://github.com/elastisys/compliantkubernetes)
- Is this changeset backwards compatible for existing clusters? Applying:
    - [ ] is completely transparent, will not impact the workload in any way.
    - [x] requires running a migration script.
    - [ ] will create noticeable cluster degradation.
          E.g. logs or metrics are not being collected or Kubernetes API server
          will not be responding while upgrading.
    - [ ] requires draining and/or replacing nodes.
    - [ ] will change any APIs.
          E.g. removes or changes any CK8S config options or Kubernetes APIs.
    - [ ] will break the cluster.
          I.e. full cluster migration is required.

### CODE REVIEW

File: helmfile/charts/harbor-backup/templates/harbor-backup-job.yaml [hunk 1]
```diff
@@ -15,9 +15,12 @@ spec:
           restartPolicy: OnFailure
           containers:
             - name: run
-              image: elastisys/backup-postgres:1.0.0
+              image: elastisys/backup-postgres:1.1.0
```
Comment (Xartos) [inline helmfile/charts/harbor-backup/templates/harbor-backup-job.yaml:18]: Will need to hold this PR until this is available. (should be soon)

**Review by lentzi90 — COMMENTED** [review]
This looks good to me!
However, I'm not sure how to test it so if you want me to do that I will need some help.
One thing in particular that we should make sure is that it is possible to push to harbor. This has been broken before due to different object storage implementations and "redirect support".

**Comment by Xartos** [comment]
> This looks good to me!
> However, I'm not sure how to test it so if you want me to do that I will need some help.
> One thing in particular that we should make sure is that it is possible to push to harbor. This has been broken before due to different object storage implementations and "redirect support".

It would be great if someone else tested this as well.
You'll need to create a new service-account in a google project. Give it storage admin rights (it was some service that failed if you didn't) and create a json key for it (the keyfile). And then paste the content of that file in the secrets file under
`objectStorage.gcs.keyFileData`.
Then create the buckets that is specified in the <sc/wc>-config files. And that should be it

File: helmfile/values/opendistro-es.yaml.gotmpl [hunk 2]
```diff
@@ -14,23 +14,44 @@ elasticsearch:
...
-    secretKey: {{ .Values.s3.secretKey }}
-    bucketName: {{ .Values.s3.buckets.elasticsearch }}
+    accessKey: {{ .Values.objectStorage.s3.accessKey }}
+    secretKey: {{ .Values.objectStorage.s3.secretKey }}
     provider: {{ .Values.global.cloudProvider }}
+    bucketName: {{ .Values.objectStorage.buckets.elasticsearch }}
+  {{ else if eq .Values.objectStorage.type "gcs" -}}
+  # TODO
+  # Add possibility to set key file data here
+  # Checkout https://github.com/elastisys/opendistro-build
```
Comment (viktor-f) [inline helmfile/values/opendistro-es.yaml.gotmpl:29]: This can be removed right?
```suggestion
```

**Comment by lentzi90** [comment]
> It would be great if someone else tested this as well.
> You'll need to create a new service-account in a google project. Give it storage admin rights (it was some service that failed if you didn't) and create a json key for it (the keyfile). And then paste the content of that file in the secrets file under
> `objectStorage.gcs.keyFileData`.
> Then create the buckets that is specified in the <sc/wc>-config files. And that should be it

I can give it a go! But I suspect that I won't have time to finish it today. Considering how much changed since I last tried to get a cluster up I will probably run into some issues that will push it to tomorrow. :sweat_smile:

**Review by viktor-f — COMMENTED** [review]
Overall it looks nice, but I agree that someone should test it with both google and s3. Not sure that I will have time for that though.

File: config/secrets/wc-secrets.yaml [hunk 3]
```diff
@@ -1,6 +1,4 @@
-s3:
-  accessKey:
-  secretKey:
+objectStorage: {}
```
[the remark this answers is not in the document]
Comment (Xartos) [reply config/secrets/wc-secrets.yaml:1]: Hehe, woops. This was my intention, so great catch!

**Review by lentzi90 — APPROVED** [review]
Alright, I finally managed to test it!
I have checked
- harbor (pushing images)
- influxdb backups works
- fluentd-aggregator (logs end up in bucket)
- elasticsearch backups

Note that I didn't try to restore from backups.

**Comment by OlleLarsson** [comment]
> Edit: Did not try S3, should I check that as well?

Yes, please :D

**Comment by lentzi90** [comment]
Tested the same as before for S3 now as well. :heavy_check_mark: 

Totally unrelated though, I happened to see this in the fluentd aggregator logs:
```
2020-11-25 13:40:54 +0000 [warn]: #0 'force_path_style' parameter is deprecated: S3 will drop path style API in 2020: See https://aws.amazon.com/blogs/aws/amazon-s3-path-deprecation-plan-the-rest-of-the-story/
```
Do you think we need an issue for this?

**Comment by OlleLarsson** [comment]
> Tested the same as before for S3 now as well. 
> 
> Totally unrelated though, I happened to see this in the fluentd aggregator logs:
> 
> ```
> 2020-11-25 13:40:54 +0000 [warn]: #0 'force_path_style' parameter is deprecated: S3 will drop path style API in 2020: See https://aws.amazon.com/blogs/aws/amazon-s3-path-deprecation-plan-the-rest-of-the-story/
> ```
> 
> Do you think we need an issue for this?

Uhm, yeah I remember that one as well.
I think that it is only relevant for AWS S3 but not for other cloud providers for the time being at least.
But sure, we can create an issue tracking the usage of path_style access.

File: helmfile/values/opendistro-es.yaml.gotmpl [hunk 4]
```diff
@@ -39,17 +57,23 @@ elasticsearch:
...
   extraInitContainers:
-  - name: install-s3
+  - name: install-object-storage-plugin
     image: amazon/opendistro-for-elasticsearch:1.10.1
     command:
     - sh
     - -c
     - |
+      {{ if eq .Values.objectStorage.type "s3" -}}
       /usr/share/elasticsearch/bin/elasticsearch-plugin install --batch repository-s3
```
Comment (OlleLarsson) [inline helmfile/values/opendistro-es.yaml.gotmpl:67]: Should we add `.. || exit 1` here as well?

**Review by OlleLarsson — APPROVED** [review]

File: helmfile/values/opendistro-es.yaml.gotmpl (hunk 4 above)
Comment (Xartos) [reply helmfile/values/opendistro-es.yaml.gotmpl:67]: Yes, nice catch! I will sneak that in when I squash the commits

**Pull request merged**